### PR TITLE
Adding methods for zero order reactive separators

### PR DESCRIPTION
**Pull request by andrewlee94**
## Summary/Motivation:
The PR adds a new helper method for creating zero-order mass balances for systems involving both reactions and separations.

## Changes proposed in this PR:
- New `zero_order_sido_reactive module
- Updates documentation

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: watertap/core/zero_order_sido_reactive.py [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+        units=pyunits.dimensionless,
+        bounds=(1E-8, 1.0000001),
+        doc='Mass recovery fraction of water in the treated stream')
+    self.removal_frac_mass_solute = Var(
+        self.flowsheet().time,
+        self.config.property_package.solute_set,
+        domain=NonNegativeReals,
+        initialize=0.01,
+        units=pyunits.dimensionless,
+        doc='Solute removal fraction on a mass basis')
```
Comment (TimBartholomew) [inline watertap/core/zero_order_sido_reactive.py:121]: I think it would be best to have `removal_frac_mass_comp` rather than solute. This would include water. Then we can just have a constraint that enforces `removal_frac_mass_comp[t, 'H2O'] = 1 - recovery_frac_mass_H2O[t]` and do all of components mass balances with `removal_frac_mass_comp`.

This would be more general and I think better aligns with AMO in that several of the components wouldn't really be considered solutes.

I also don't see the harm in doing this for the WaterTAP3 type models. What value do we get out of the components being solutes?

File: watertap/core/zero_order_sido_reactive.py [hunk 2]
```diff
@@ -0,0 +1,416 @@
...
+                f"{self.name} - database provided does not "
+                f"contain an entry for key_reactant for reaction {r}.")
+        if key_reactant not in b.properties_in[0].component_list:
+            raise ValueError(
+                f"{self.name} - key_reactant {key_reactant} for reaction {r} "
+                f"is not in the component list used by the assigned property "
+                f"package.")
+        return (b.extent_of_reaction[t, r] ==
+                b.reaction_conversion[t, r] *
+                b.properties_in[t].flow_mass_comp[key_reactant])
```
Comment (TimBartholomew) [inline watertap/core/zero_order_sido_reactive.py:167]: Isn't this not technically the extent_of_reaction? This is the amount of the key reagent reacted. I like how the model is formulated, but I think we should rename the variable to not cause confusion.

File: watertap/core/zero_order_sido_reactive.py [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+
+    self.generation_ratio = Param(
+        self.reaction_set,
+        self.config.property_package.component_list,
+        initialize=0,
+        mutable=True,
+        units=pyunits.dimensionless,
+        doc="Mass ratio for generation of species w.r.t. key species")
```
Comment (TimBartholomew) [inline watertap/core/zero_order_sido_reactive.py:175]: Is there a reason you chose `generation_ratio` rather than `conversion_ratio`? The temporary variable you use to calculate this is `cratio` and the comments refer to it as the conversion ratio.

Comment (TimBartholomew) [reply watertap/core/zero_order_sido_reactive.py:175]: Why did you decide to go with a param here? Isn't the IDAES convention to just use fix vars?

File: watertap/core/zero_order_sido_reactive.py [hunk 4]
```diff
@@ -0,0 +1,416 @@
...
+                b.properties_byproduct[t].flow_mass_comp["H2O"])
+
+    # Solute removal
+    @self.Constraint(self.flowsheet().time,
+                     self.config.property_package.solute_set,
+                     doc='Solute removal equations')
+    def solute_removal_equation(b, t, j):
+        return (b.removal_frac_mass_solute[t, j] *
+                (b.properties_in[t].flow_mass_comp[j] +
+                 sum(b.generation_ratio[r, j]*b.extent_of_reaction[t, r]
```
Comment (TimBartholomew) [inline watertap/core/zero_order_sido_reactive.py]: I think it would be helpful to the users if we actually made a generation term (i.e. indexed variable and constraint) rather than just using the product of the generation_ratio and extent of reaction. I could see users frequently wanting to just check how much of a component was generated.

File: watertap/core/zero_order_sido_reactive.py [hunk 5]
```diff
@@ -0,0 +1,416 @@
...
+                            f"{self.name} - database provided does not "
+                            f"contain an entry for molecular_weight w.r.t. "
+                            f"species {j} in reaction {r}.")
+                    try:
+                        MW_k = cratio = stoich[key_reactant]["molecular_weight"]
+                    except KeyError:
+                        raise KeyError(
+                            f"{self.name} - database provided does not "
+                            f"contain an entry for molecular_weight w.r.t. "
+                            f"species {key_reactant} in reaction {r}.")
```
Comment (TimBartholomew) [inline watertap/core/zero_order_sido_reactive.py]: remember to remove these repeated `cratio =`

File: watertap/core/zero_order_sido_reactive.py (hunk 2 above)
Comment (TimBartholomew) [reply watertap/core/zero_order_sido_reactive.py:167]: In another comment, I suggest creating a variable for the generation term, we could just use that with the key reagent index instead of this variable.

**Review by TimBartholomew — COMMENTED** [review]
A few comments, could be faster to just have a call about them.

File: watertap/core/zero_order_sido_reactive.py (hunk 2 above)
Comment (andrewlee94) [reply watertap/core/zero_order_sido_reactive.py:167]: 1. Extent of reaction is always relative to something, and normally it is expressed relative to the key component so I would say that yes this is the extent of reaction.
2. Rather than a Var for generation I would suggest an Expression if you think it is important. There is no need to introduce additional Vars and Constraints into the problem for this, as that is just making the problem bigger than it needs to be.

File: watertap/core/zero_order_sido_reactive.py (hunk 3 above)
Comment (andrewlee94) [reply watertap/core/zero_order_sido_reactive.py:175]: 1. I think the two terms are essentially synonymous, and that is just what I picked.
2. No, the IDAES convention is to use Vars for anything that might conceivably be solved for, however this is something that definitely should not be left free. Hence the use of a Param.

File: watertap/core/zero_order_sido_reactive.py (hunk 1 above)
Comment (andrewlee94) [reply watertap/core/zero_order_sido_reactive.py:121]: I strongly suggest not including water in here for two reasons.

1. It adds the possibility of over-specifying the problem as a user could specify both the removal and recovery of water (or worse yet, specify both of those and leave another removal unfixed and wonder why their problem with 0 degrees of freedom does not solve). Believe me, someone will try to do this. On top of this, think of the additional logic that would need to be implemented to work out which of these to fix when loading parameters from a datafile. What do we do if a user provides values for both recovery and removal?
4. It adds an additional Var and Constraint to the problem that are not really required. We want to keep these models as simple as possible, so I don't see much value in adding an additional Var and Constraint for this.

On the comment about naming, that is primarily semantics. At the moment, "solute" just means anything that is not water, and yes, not everything in there is strictly a solute (it also includes suspended solids, other solvents and potentially entrained gases). However, going with the full component list is not the solution to this. At most, we should use a more inclusive name (but I have no idea what that is). Further, if we do that then we need to go change everything else in the zero-order models too.

Comment (TimBartholomew) [reply watertap/core/zero_order_sido_reactive.py:121]: For comment 1, this is the case in all models. Users could fix variables that they shouldn't, and the model isn't solvable but still have zero degrees of freedom. The most similar model to these is the IDAES separator model, there users have the option to fix split fractions for either outlet stream. If they fix incompatible split fraction for one component in both outlet streams and omit a split fraction for another component, the model has zero degrees of freedom and is not solvable.

For comment 2, I understand the reasoning for not wanting another variable and constraint, especially for introducing a bilinear term into the mass balance for water. An alternative would be to fix the removal fraction of water given the recovery (rather than a constraint), i.e. we just make model only consider removal fractions and drop the recovery (in this case everyone knows they need to supply the removal fraction for water, which is 1 - recovery).

It just seems like an odd choice to make all the components solutes when they aren't and that the variables names explicitly include solute in them. It also makes the water balance different than all the other components when it is not really necessary.

For example, one AMO project has a reactor that creates a product gas with negligible water content. Then that product gas is sent to a membrane unit which should be modeled as zero order. If I use our zero order models, it's weird the removal fraction has solute in the name. It's also weird that there is a separate water balance and a requirement to provide its recovery, it would make more sense if the removal fraction of water was just automatically populated with the default value since it doesn't matter.

In summary, I feel like differentiating water from all other components is adding an additional layer of complexity, when really it can be treated the same. Then making all other components a solute is an unnecessary restriction that just is odd in several cases.

**Review by TimBartholomew — APPROVED** [review]

File: watertap/core/zero_order_sido_reactive.py (hunk 2 above)
Comment (adam-a-a) [reply watertap/core/zero_order_sido_reactive.py:167]: I was a little confused by this as well. Following the definition of extent of reaction, represented by delta_n/stoich #, and working out how you formulated things, I think the extent of reaction should be  
`extent_of_reaction == reaction_conversion * flow_mass_comp[key_reactant] / MW[key_reactant]` which should work out to be equivalent to moles/s generated divided by stoichiometric number.

Comment (andrewlee94) [reply watertap/core/zero_order_sido_reactive.py:167]: The problem is that we will not always have a MW for the components (we explicitly allow for empirical mass conversion ratios), hence we can't do this on a molar basis. Thus, this is a mass-based extent of reaction instead.

File: watertap/core/zero_order_sido_reactive.py [hunk 6]
```diff
@@ -0,0 +1,424 @@
...
+                b.properties_byproduct[t].flow_mass_comp["H2O"])
+
+    # Solute removal
+    @self.Constraint(self.flowsheet().time,
+                     self.config.property_package.solute_set,
+                     doc='Solute removal equations')
+    def solute_removal_equation(b, t, j):
+        return (b.removal_frac_mass_solute[t, j] *
+                (b.properties_in[t].flow_mass_comp[j] +
+                 sum(b.generation_rxn_comp[t, r, j]
```
Comment (adam-a-a) [inline watertap/core/zero_order_sido_reactive.py:264]: While this formulation works, I am thinking about cases where this formulation may not be applicable (or easily applicable) such as in biological processes where Monod kinetics would be used to model microbial growth. Thus, I am wondering if there should be a base class that includes the `generation_rxn_comp` term, and then this formulation would inherit the specifics that arrive at defining that generation term. Subsequently, as different representations for the generation term might be useful or desired, we could inherit said base class and add the necessary specifics to define that generation term. I hope that makes sense.

Comment (adam-a-a) [reply watertap/core/zero_order_sido_reactive.py:264]: I just backtracked and recalled that this is not a class, so I guess that suggestion isn't directly applicable. Still something to think about -- we _may_ need alternative representations for generation/decay.

Comment (andrewlee94) [reply watertap/core/zero_order_sido_reactive.py:264]: That is in part why we didn't insist on stoichiometric reactions, and left things fairly empirical.

File: watertap/core/zero_order_sido_reactive.py (hunk 1 above)
Comment (adam-a-a) [reply watertap/core/zero_order_sido_reactive.py:121]: Just my two cents- while I agree with @TimBartholomew that we don't _need_ to differentiate water from all other components, I think that we don't have time to spare to step back and revise everything that is already in place for WaterTAP3-related models because it would set us back. Instead, I think our time would be better spent by pushing forward and get all models implemented first. So I have to agree with @andrewlee94 on this one. With regard to not all "solutes" necessarily being "solutes", I suppose a more appropriate name would be something like "constituent", but again, since time is limited and altering the name across the board wouldn't change code functionality, I think we can live with what is in place now. Once all of our models are in place and we shift our focus towards refinement, we could go back and refine as needed.

File: watertap/core/zero_order_sido_reactive.py (hunk 6 above)
Comment (andrewlee94) [reply watertap/core/zero_order_sido_reactive.py:264]: Also, kinetics don't belong in zero-order models.

**Review by adam-a-a — APPROVED** [review]
I think this is good to go. I had a couple of comments to consider but nothing that should hold up the PR.
